core/execution/spot/equity_shares.go

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2022 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE.VEGA file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package spot
+
+import (
+	"fmt"
+
+	"code.vegaprotocol.io/vega/libs/num"
+)
+
+// lp holds LiquidityProvider stake and avg values.
+type lp struct {
+	// physical stake
+	buyStake      num.Decimal
+	sellStake     num.Decimal
+	physicalStake num.Decimal
+	// virtual stake
+	buyVStake    num.Decimal
+	sellVStake   num.Decimal
+	virtualStake num.Decimal
+
+	share num.Decimal
+	avg   num.Decimal
+}
+
+type EquityShares struct {
+	// market value proxy
+	mvp num.Decimal
+	// growth factor
+	r num.Decimal
+
+	totalVStake num.Decimal
+	totalPStake num.Decimal
+	lps         map[string]*lp
+
+	openingAuctionEnded bool
+}
+
+func NewEquityShares(mvp num.Decimal) *EquityShares {
+	return &EquityShares{
+		mvp:         mvp,
+		r:           num.DecimalZero(),
+		totalPStake: num.DecimalZero(),
+		totalVStake: num.DecimalZero(),
+		lps:         map[string]*lp{},
+	}
+}
+
+// OpeningAuctionEnded signal to the EquityShare that the
+// opening auction has ended.
+func (es *EquityShares) OpeningAuctionEnded() {
+	// we should never call this twice
+	if es.openingAuctionEnded {
+		panic("market already left opening auction")
+	}
+	es.openingAuctionEnded = true
+	es.r = num.DecimalZero()
+}
+
+func (es *EquityShares) updateVStake(markPrice num.Decimal) {
+	if es.r.IsZero() {
+		return
+	}
+	total := num.DecimalZero()
+	factor := num.DecimalOne().Add(es.r)
+	recalc := false
+	for _, v := range es.lps {
+		v.buyVStake = num.MaxD(v.buyStake, v.buyVStake.Mul(factor))
+		v.sellVStake = num.MaxD(v.sellStake, v.sellVStake.Mul(factor))
+		v.virtualStake = num.MinD(v.buyVStake, v.sellVStake.Mul(markPrice))
```

Comment:
Wouldn't this almost always guarantee that the virtual stake == sell virtual stake (except for when the sell virtual stake == 0). The buy virtual stake is either the physical stake or physical * (1 + r). With a mark price of, say, 10,000, the market growth would have to be insane, and even then, the virtual stake is scaled using the same growth factor.

Reply:
I don't follow. multiplying by mark price is to convert from base to quote, if the commitment is equal (when converting to quote space), there's no reason why one would be by definition greater than the other. Maybe I'm missing your point?

The way the above works is that the virtual stake for each is managed in the native token, so in quote for buy and in base for sell, then when calculating the virtual stake we choose the lower between the buy/sell virtual stake, having converted the sell to quote space.

Reply:
Ignore, just reading through the PR thinking about futures, not spot markets